pkg/serverhandler/handlertest/test_service_ui_event.go

```diff
@@ -0,0 +1,157 @@
+package handlertest
+
+import (
+	"context"
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+	serverptypes "github.com/hashicorp/waypoint/pkg/server/ptypes"
+	"github.com/stretchr/testify/require"
+	"testing"
+	"time"
+)
+
+func init() {
+	tests["event"] = []testFunc{
+		TestEvent,
+	}
+}
+func TestEvent(t *testing.T, factory Factory) {
+	ctx := context.Background()
+
+	// Create our server
+	client, _ := factory(t)
+
+	t.Run("paginate events", func(t *testing.T) {
+		require := require.New(t)
+
+		proj := &pb.Project{
+			Name: "p_test",
+		}
+		refProj := &pb.Ref_Project{Project: "p_test"}
+		_, err := client.UpsertProject(context.Background(), &pb.UpsertProjectRequest{
+			Project: proj,
+		})
+		require.NoError(err)
+
+		_, err = client.UpsertApplication(context.Background(), &pb.UpsertApplicationRequest{
+			Project: refProj,
+			Name:    "a_test",
+		})
+		require.NoError(err)
+
+		// Create build, deployment, release
+		createEvents(client, ctx, t)
+
+		time.Sleep(250 * time.Millisecond)
```

Comment:
Curious: what is the sleep here accomplishing?

Reply:
My idea for the `time.Sleep` was to have a different Status Start & Complete time for the created resources.

Reply:
My understanding is that we emit 2 events for the same operation, one with status.startTime filled in, and one with both status.startTime + status.completeTime filled in. Is that correct?

At the state layer, we then upsert events so the same event gets updated with a completeTime when it's complete?

On the frontend then, depending on when we call UI_ListEvents, we could get an event with either a startTime but no completeTime or both startTime and completeTime, but we would always only get one event per operation?

Reply:
I believe that is correct. As UI_ListEvents calls the EventListBundles function which will list the events from the db. An event will have a completed time in the status field only after its completed and the second Upsert<Operation> call is made from the runner.

Each Upsert<Build/Deploy/Release> gets called upon the correlative cli commands, that puts the operation in the db, then the runner will call it again to update that record to put in the completed time once its done. EventPut operates in the same way as it gets called in the Upsert<Operation>. This will put an event the first time it is 'started', then once the runner updates the operation record the second time it calls Upsert<Operation> it will update the correlative event record with a new EventTimestamp and eventData containing that completed time.